src/llfs/page_cache_slot_pinned_ref.hpp

```diff
@@ -12,6 +12,26 @@
 
 namespace llfs {
 
+//==#==========+==+=+=++=+++++++++++-+-+--+----- --- -- -  -  -   -
+//
+class CallerPromisesTheyAcquiredPinCount
```

Comment:
The usage of this class made me think of `PageCacheSlot::extend_pin` as well. Since `extend_pin` is called in a slightly different context than `acquire_pin` and `fill` and also doesn't construct a `PinnedRef`, I understand that the comparison isn't directly analogous. I was wondering if we'd want to similarly enforce some sort of control on this operation to ensure correctness in where it can be called? Just thinking out loud since it's another operation that manipulates pin count, but I do see that the only place it seems to be called for now is the `PinnedRef` copy constructor.

Reply:
I think `extend_pin` is a little different because we can pretty much verify all the pre-conditions from inside the function (i.e., `extend_pin` can **never** take the pin count from 0 to 1).  `CallerPromisesTheyAcquiredPinCount`, in contrast, locks down the places in the code that are allowed to construct a `PinnedRef` _ex-nihilio_, to reduce the probability that one of us writes code in the future that constructs a PinnedRef _without_ having acquired a unique pin count previously.

Reply:
Makes sense, thanks for this explanation!